kaal-infrastructure/src/main/kotlin/cz/eman/kaal/infrastructure/api/retrofit/KaalRetrofitCaller.kt

```diff
@@ -2,127 +2,210 @@ package cz.eman.kaal.infrastructure.api.retrofit
+        return callResultNull(responseCall, errorMessage, map).map { data ->
+            if (data != null) {
+                Optional.of(data)
+            } else {
+                Optional.empty()
+            }
+        }
+    }
 
     /**
      * Calls a [responseCall] and handles the result by mapping the [Dto] object to nullable [T] on success else it
      * returns [Result.Error] on any error or exception. Optional allows support for empty body.
      *
+     * Handles response using [handleResponse] and is returned. If any exception occurs it is handled by
+     * [handleCallException].
+     *
      * @param responseCall Retrofit2 call to handle
      * @param errorMessage used to modify error message
      * @param map function mapping [Dto] object to [T] object
      * @return [Result] with [T]? (nullable) allowing success with empty body
+     * @see responseCall
+     * @see handleResponse
+     * @see handleCallException
      * @since 0.9.0
      */
     suspend fun <Dto, T> callResultNull(
         responseCall: suspend () -> Response<Dto>,
         errorMessage: () -> String?,
         map: suspend (Dto) -> T
-    ): Result<T?>
+    ): Result<T?> {
+        return try {
+            handleResponse(responseCall(), map)
+        } catch (ex: Exception) {
+            handleCallException(ex, errorMessage())
+        }
+    }
 
     /**
      * Calls a [responseCall] and handles the result by mapping the [Dto] object to [Optional]<[T]> on success else it
      * returns [Result.Error] on any error or exception. Optional allows support for empty body. Result of this function
      * is a [Pair] of [Result] and [Response] so you can easily access any response metadata you need (like headers,
      * result code and others provided by Retrofit2).
      *
+     * This function calls [callResultCompleteNull] which can return [Result] result with null body. This result is then
+     * mapped to [Optional] using [Result.map].
+     *
      * @param responseCall Retrofit2 call to handle
      * @param errorMessage used to modify error message
      * @param map function mapping [Dto] object to [T] object
-     * @return [Pair] of [Result] with [T] data to [Response] with [Dto] object
+     * @return [CompleteRetrofitResponse] with [T] (Optional) and [Dto] object
+     * @see responseCall
+     * @see handleResponse
+     * @see handleCallException
      * @since 0.9.0
      */
     @RequiresApi(Build.VERSION_CODES.N)
-    suspend fun <Dto, T> callResultResponseOptional(
+    suspend fun <Dto, T> callResultCompleteOptional(
         responseCall: suspend () -> Response<Dto>,
         errorMessage: () -> String?,
         map: suspend (Dto) -> T
-    ): Pair<Result<Optional<T>>, Response<Dto>?>
+    ): CompleteRetrofitResponse<Optional<T>, Dto> {
+        val response = callResultCompleteNull(responseCall, errorMessage, map)
+        val mappedResult = response.result.map { data ->
+            if (data != null) {
```

Comment:
the same construct has been used for response of callResultNull in CallResultUnit method, i recommend to extract this mapper to separated method